processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java

```diff
@@ -760,12 +761,22 @@ public ConstantMappingBuilder selectionParameters(SelectionParameters selectionP
         public PropertyMapping build() {
             // source
             String sourceErrorMessagePart = "constant '" + constantExpression + "'";
+            String errorMessageDetails = null;
 
-            Type sourceType = ctx.getTypeFactory().getTypeForConstant( targetType, constantExpression );
-            if ( String.class.getCanonicalName().equals( sourceType.getFullyQualifiedName() ) ) {
-                // convert to string
+            Class<?> baseForLiteral = null;
+            try {
+                baseForLiteral = NativeTypes.getLiteral( targetType.getFullyQualifiedName(), constantExpression );
+            }
+            catch ( IllegalArgumentException ex ) {
+                errorMessageDetails = ex.getMessage();
+            }
+
+            // asume a String constant
```

Comment:
typo, asume -> assume
Not sure of mapstructs standards, but when I write a comment, I try to explain stuff in some detail. The comment exists, it should justify itself. Here, I'd probably write something like:
`// the constant is not a primitive literal, assume it to be a String`

Reply:
I'll adapt.